### DISCUSSION TITLE
[Proposal] Annotation Blocks

### DISCUSSION THREAD

**Discussion by sirgru** [dotnet/csharplang#818] [General]
### Motivation ###
Currently attributes and access modifiers have to be specified for all members individually. This can lead to ugly and overly verbose declarations. Further, in VS attributes will be auto-corrected to be on a different line to a declaration, which makes it difficult to have attributes inline.
```C#
[Atr1]
[Atr2]
public MyType1 myVar1 {get; private set;}
[Atr1]
[Atr2]
public MyType2 myVar2 {get; private set;}
[Atr1]
[Atr2]
public MyType3 myVar3 {get; private set;}
[Atr1]
[Atr2]
public MyType4 myVar4 {get; private set;}
```

### Solution ###
Allow blocks, somewhat similar to C++ access modifier blocks, with syntax like follows.
```C#
[Atr1] [Atr2] public : {
    MyType1 myVar1
    MyType2 myVar2
    MyType3 myVar3
    MyType4 myVar4
}: {get; private set;}
```
They have the following features:
1. Start with attributes or access modifiers we like to share, followed by `:{`
2. Can end with `}` or `}:`. The first means no suffixes and second means there is a suffix following, like accessors or default values.
3. Compiler should know prefixes and postfixes are applied to every line. This should reflect in Intellisense.

**Comment by svick** [comment]
I think this makes code easier to write, but harder to read, which is not a good trade-off most of the time.

> in VS attributes will be auto-corrected to be on a different line to a declaration, which makes it difficult to have attributes inline

If you don't like how VS behaves, that's a reason to change how VS behaves, not change the whole language. Also, using the `[Atr1, Atr2]` syntax could help.

**Comment by iam3yal** [comment]
A case for code generator?

Something like:

```
[Atr1]
[Atr2]
public MyType1 myVar1 { get; private set; }

[CopyAttributes(nameof(myVar1))]
public MyType2 myVar2 { get; private set; }

[CopyAttributes(nameof(myVar1))]
public MyType3 myVar3 { get; private set; }

[CopyAttributes(nameof(myVar1))]
public MyType4 myVar4 { get; private set; }
```

**Comment by sirgru** [comment]
@svick As far as VS, that's a fair point, but it was an added motivation for the proposal for me. Even with one attribute it gets quite annoying, but that's an issue for another repo.

As far as readability, that can be arguable. I find the separation much easier to read in my example. The presence of indentation could clearly show there's a code block going on even if the number of declarations is greater than the height of the screen. Now long attributes can cover variable names which makes the whole space cluttered when looking through our own code. There is a situation when a certain attribute is used quite often, and that just becomes visual clutter.

@eyalsk Still lots of copy-pasting going on :) It's code duplication!

**Comment by iam3yal** [comment]
@sirgru 

> @eyalsk Still lots of copy-pasting going on :) It's code duplication!

```
[Atr1, Atr2]
[ApplyClassAttributesToProperties]
public class Foo
{
	public MyType1 myVar1 { get; private set; }
	
	public MyType2 myVar2 { get; private set; }
	
	public MyType3 myVar3 { get; private set; }
	
	public MyType4 myVar4 { get; private set; }
}
```

Yes, it isn't terse as your example but it's probably more readable. ;)

p.s. I'd argue that code duplication as in text isn't bad (and we're doing it all the time), it's duplication of knowledge that is bad.

**Comment by Joe4evr** [comment]
That would be a compiler error if `[Atr1]` and `[Atr2]` didn't have `AttributeTargets.Class` specified in their `[AttributeUsage]`.

**Comment by iam3yal** [comment]
@Joe4evr Why would it be a compile-time error? they are derived from `Attribute` that has `AttributeTargets.All` set so if you don't set any restriction it should compile just fine.

**Comment by Joe4evr** [comment]
> so if you don't set any restriction it should compile just fine.

Well yeah, that's the key thing, but I'm pretty sure a good majority of well-used attributes *are* restricted. Also, it just feels plain dirty to create an attribute class that isn't decorated with `[AttributeUsage]`. Even re-declaring `AttributeTargets.All` is something I would highly recommend so that it shows up in VS's Metadata as Source, but I digress.

**Comment by iam3yal** [comment]
@Joe4evr Do you realize this is an example? I really don't get what's the point of this.. you went from saying it won't compile to I don't like the code..

**Comment by ghost** [comment]
"C++ access modifier blocks" ranks high on my list of things that I despise whenever I have to read  C++ code. I like having the visibility information right there on the method definition where I don't have to go hunting for it. This syntax would open the door to that - for that alone, it gets a thumbs-down from me.

**Comment by sirgru** [comment]
@AtsushiKan I can imagine it being pretty bad with methods, I thought only of fields / properties.

**Comment by Thaina** [comment]
Inversely

```C#
[Atr1]
[Atr2]
[CopyAttributesTo(nameof(myVar2),nameof(myVar3),nameof(myVar4))]
[Attr3] // not copied
public MyType1 myVar1 { get; private set; }
public MyType2 myVar2 { get; private set; }
public MyType3 myVar3 { get; private set; }
public MyType4 myVar4 { get; private set; }
```
